### ISSUE TITLE
Project Invariants

### ISSUE THREAD

**Issue by dbramucci** [dkhwang118/CIS560_F18-Project-RecipeManager#12]
List properties that should always be true before and after a method in our program.

**Comment by dbramucci** [comment]
A quantity should never be negative because you can't have -1 ounces of salt.

This involves checking for sufficient amounts of material prior to subtracting out material.

**Comment by dbramucci** [comment]
A recipe, pantry or shopping list should never refer to an ingredient that doesn't exist.

This should be enforced by database constraints and gui design.

**Comment by laurel-thomson** [comment]
> A recipe, pantry or shopping list should never refer to an ingredient that doesn't exist.
> 
> This should be enforced by database constraints and gui design.

Maybe to start with, we should not allow users to delete Ingredients?  I can see enforcing this invariant being a little tricky if we allow the user to delete Ingredients.  We would need to first check to see if the Ingredients are used in any Recipes and ShoppingLists.  An Ingredient in a ShoppingList is a bit simpler: we could just remove it from the ShoppingList.  But removing an Ingredient from a Recipe would mess up the Recipe and we'd want to give the user a chance to replace it with something.

**Comment by dkhwang118** [comment]
> > A recipe, pantry or shopping list should never refer to an ingredient that doesn't exist.
> > This should be enforced by database constraints and gui design.
> 
> Maybe to start with, we should not allow users to delete Ingredients? I can see enforcing this invariant being a little tricky if we allow the user to delete Ingredients. We would need to first check to see if the Ingredients are used in any Recipes and ShoppingLists. An Ingredient in a ShoppingList is a bit simpler: we could just remove it from the ShoppingList. But removing an Ingredient from a Recipe would mess up the Recipe and we'd want to give the user a chance to replace it with something.

I agree that for now we should not allow users to delete/update ingredients. Add and Search should be the focus while we build things up in the beginning.

When we arrive at the issue later we can do a few different things to allow the user to update/delete ingredients, but I believe that when we get to that stage we'll also have to decide on how to update/delete recipes, shopping lists, etc. so we can standardize the way we handle update/delete from the database.

**Comment by montgomerysamantha** [comment]
I'm not sure if I'm understanding this correctly, but wouldn't we want to add new Ingredients to the database if they were mentioned in a Recipe or ShoppingList? For example, a user puts in a new Recipe with a couple of new Ingredients. So, the Recipe would be referencing a nonexistent Ingredient. For simplicity, I think we would add the new Ingredient(s) to the database. It would be a big hassle for the user to have to add every new Ingredient before adding the full Recipe.

Or maybe I'm missing something. Let me know.

**Comment by laurel-thomson** [comment]
It sounds like for now, we will only allow users to add Ingredients and not allow them to update or delete Ingredients.  As far as the UI design for whoever works on #3, I would agree that we want the user to be able to add new Ingredients as they are creating a Recipe (not force them to first add the Ingredient in the Pantry form).
